lib/core/src/Cardano/Wallet.hs

```diff
@@ -765,26 +764,30 @@ newWalletLayer tracer bp db nw tl = do
         -> Maybe UTCTime
         -> SortOrder
         -> ExceptT ErrListTransactions IO [TransactionInfo]
-    _listTransactions wid mStart mEnd order = do
-        guardRange (mStart, mEnd)
-        (w, _) <- withExceptT ErrListTransactionsNoSuchWallet $ _readWallet wid
-        let tip = currentTip w ^. #slotId
-        let range = Range
-                { rStart = slotCeiling sp <$> mStart
-                , rEnd = slotFloor sp <$> mEnd
-                }
-        liftIO $ assemble tip
-            <$> DB.readTxHistory db (PrimaryKey wid) order range
+    _listTransactions wid mStart mEnd order =
+        maybe (pure []) listTransactionsWithinRange =<< getSlotRange
```

Comment:
@Anviking wrote:

> So if the range is too low, we return `[]` directly. I slightly fear this is scattering the filtering-logic.

I agree with you here.

The core problem is that we're unable to express the _empty_ range, since both lower and upper bounds are _inclusive_.  Therefore, we must resort to tricks like `Maybe (Range a)`.

However, if we change `Range` so that it _is_ possible to express the empty range, then we can just pass the empty range into the `listTransactionsWithRange` function.

I propose a simple change to accommodate this: make the upper bound of `Range` exclusive. This way, the range `Range (Just n) (Just n)` is _empty_ for all values of `n`.

Of course, we _could_ do something more general like allowing each bound to be _either_ inclusive _or_ exclusive, but I think we can get away without doing that for the moment. If we need to make it more general later on, we can always do that.

Reply:
I think the reason we had inclusive bounds was so that clients always could fetch the next page without gaps, despite slotId/insertedAt not being unique, even if that meant that it would have to prune duplicates at the edges. (😐)

With `[a,b)` I think this would be fine for `Ascending` but not fine for `Descending`.

I think what you have in the PR is very fine, and we should stick with it unless we have an obvious alternative.

Reply:
> I think the reason we had inclusive bounds was so that clients always could fetch the next page without gaps, despite slotId/insertedAt not being unique, even if that meant that it would have to prune duplicates at the edges. (neutral_face)
> 
> With `[a,b)` I think this would be fine for `Ascending` but not fine for `Descending`.
> 
> I think what you have in the PR is very fine, and we should stick with it unless we have an obvious alternative.

Okay, in that case, I'll polish up this PR so that it's ready for final review.

We can think about a more general solution later on. :)